packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/useHandleUpload.ts

```diff
@@ -67,39 +73,70 @@ const updateTasks = <T extends Task | CancelableTask>(
   return [...tasks.slice(0, index), updatedTask, ...tasks.slice(index + 1)];
 };
 
+// Helper function to replace files with the same name
+const mergeSelectedFiles = (prevFiles: File[], newFiles: File[]): File[] => {
+  const files: File[] = [];
+  const filesSet = new Set<string>();
+
+  // Add new files so they appear on the top of the table
+  newFiles.forEach((file) => {
+    const fileKey = getFileKey(file);
+
+    files.push(file);
+    filesSet.add(fileKey);
+  });
+
+  prevFiles.forEach((file) => {
+    const fileKey = getFileKey(file);
+
+    if (!filesSet.has(fileKey)) {
+      files.push(file);
+    }
+  });
+
+  return files;
+};
+
 export function useHandleUpload({
   prefix,
-  files,
   preventOverwrite,
-  updateFiles,
 }: {
   prefix: string;
-  files: File[];
   preventOverwrite: boolean;
-  updateFiles: React.Dispatch<React.SetStateAction<File[]>>;
-}): [tasks: CancelableTask[], handleUpload: () => void] {
+}): [
+  tasks: CancelableTask[],
+  handleUpload: () => void,
+  handleFileSelect: (files: File[]) => void,
+] {
   const getConfig = useGetLocationConfig();
 
   const [tasks, setTasks] = React.useState<CancelableTask[]>(() => []);
+  const [_, setFiles] = React.useState<File[]>([]);
```

Comment:
Added a `files` state here so that I could keep track of what files have been added and remove them when the task is removed.

Would this be okay, or is this coupling the files and tasks too closely?